AX: Implement TextUnit::Paragraph off the main thread

hoffmanjoshua opened the pull request:
#### bf209d7654217b9104b9202240b934103414e7ff
<pre>
AX: Implement TextUnit::Paragraph off the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=282310">https://bugs.webkit.org/show_bug.cgi?id=282310</a>
<a href="https://rdar.apple.com/134579753">rdar://134579753</a>

Reviewed by Tyler Wilcock.

This PR implements the current paragraph API to be served from the ax thread when AX_THREAD_TEXT_APIS
is enabled. Similar to computing the current line API, paragraph borders are determined by newline
characters, renderers with different containing blocks, and line break elements.

The text-marker-paragraph-nav.html test was partially ported to verify the behavior of this API.
A follow up PR will implement the previousParagraphStart and nextParagraphEnd APIs.

* LayoutTests/accessibility/ax-thread-text-apis/test.mp3: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::paragraphRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::previousParagraphStart const):
(WebCore::AXTextMarker::nextParagraphEnd const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::startsWithLineBreak const):
(WebCore::AXTextRun::endsWithLineBreak const):
(WebCore::AXTextRuns::runLength const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):

Canonical link: <a href="https://commits.webkit.org/286004@main">https://commits.webkit.org/286004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da7e3e20a6ffd6e6b027e78d87a825e843d9850f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53741 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27123 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78692 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25550 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76429 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62874 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58413 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16742 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48582 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63918 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38825 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45563 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21413 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23883 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66973 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21760 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80208 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1629 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/937 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66706 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1777 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63936 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66009 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16416 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9928 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8077 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1593 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4381 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1622 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1610 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1629 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
-    // FIXME: Implement the other AXTextUnits.
+
+    // Paragraphs must be handled differently from word + sentence boundaries, as there is no paragraph break iterator.
+    // Rather, paragraph boundaries are based on rendered newlines and changes in nodes/renderers.
```
"changes in nodes/renderers" is a little bit vague. Is it possible to be more specific here — what changes from node-to-node or renderer-to-renderer are relevant?

hoffmanjoshua replied:
Will update this to be more specific!

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+            for (size_t i = startLineIndex; i < currentRuns->size() && i >= 0; direction == AXDirection::Next ? i++ : i--) {
+                // If a text run starts or ends with a newline character, that indicates a paragraph boundary.
+                if (currentRuns->at(i).endsWithLineBreak() && (i != startLineIndex || (direction == AXDirection::Next && currentRuns->runLength(i) < offsetInStartLine))) {
+                    unsigned newlineOffsetConsideringDirection = direction == AXDirection::Next ? cumulativeOffset + currentRuns->runLength(i) - 1 : cumulativeOffset;
```
`currentRuns->runLength(i) - 1` — subtracting one is safe because textruns should always be non-zero, right? I feel like you said that in a previous PR. Do you know if we have some `ASSERT` that checks this somewhere? e.g. maybe when we build the text runs. Or maybe we could put the `ASSERT` in `runLength` (if there isn't already an `ASSERT` somewhere) with a comment saying several places rely on this being non-zero when subtracting the result?

hoffmanjoshua replied:
Yeah, I confirmed in a previous patch that runLength can't be zero. I think an assert is a good idea—I'll go ahead and add that

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+    // Rather, paragraph boundaries are based on rendered newlines and changes in nodes/renderers.
+    if (textUnit == AXTextUnit::Paragraph) {
+        size_t startLineIndex = currentRuns->lineID(runIndex).lineIndex;
+        unsigned offsetInStartLine = currentRuns->runLengthSumTo(startLineIndex - 1) - offset();
```
Can we ever compute a `startLineIndex` of zero (thus making this `- 1` underflow)? Or is that impossible?

hoffmanjoshua replied:
Just updated and cleaned up this bit. Now, we just use the `startLineIndex` without subtracting 1, and compute the offset in a more correct way.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+            currentObject = findObjectWithRuns(*currentObject, direction, stopAtID);
+            currentRuns = currentObject ? currentObject->textRuns() : nullptr;
+
+            // Paragraph boundarys also change based on editability, containing block, and whether we hit a line break.
```
Typo: "boundarys" should be "boundaries"

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+
+            // Paragraph boundarys also change based on editability, containing block, and whether we hit a line break.
+            bool isContainingBlockBoundary = currentRuns && previousRuns && currentRuns->containingBlock != previousRuns->containingBlock;
+            bool isEditBoundary = previousObject && currentObject && !!previousObject->editableAncestor() != !!currentObject->editableAncestor();
```
We don't need to compute `isEditBoundary` if `isContainingBlockBoundary`. Since the latter does two ancestry walks, maybe we could do this with a comment making it clear why we do it, even if it's a bit verbose:

```cpp
bool isContainingBlockBoundary = ...;
// Don't bother computing isEditBoundary if isContainingBlockBoundary since we only need one or the other below.
bool isEditBoundary = !isContainingBlockBoundary && previousObject && ...;
```

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4762,7 +4764,7 @@ AXTreeData AXObjectCache::treeData()
         stream << "\nAXIsolatedTree:\n";
         if (auto tree = getOrCreateIsolatedTree()) {
-            constexpr OptionSet<AXStreamOptions> options = { AXStreamOptions::ObjectID, AXStreamOptions::ParentID };
+            constexpr OptionSet<AXStreamOptions> options = { AXStreamOptions::ObjectID, AXStreamOptions::ParentID, AXStreamOptions::TextRuns };
```
Do we want to commit this change?

hoffmanjoshua replied:
Ah no this was inadvertently included, but, maybe we should add this by default when `AX_THREAD_TEXT_APIS` is true? That can be for another day though...

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -3758,8 +3758,10 @@ CharacterOffset AXObjectCache::previousBoundary(const CharacterOffset& character
+// MARK: replicate this!
 CharacterOffset AXObjectCache::startCharacterOffsetOfParagraph(const CharacterOffset& characterOffset, EditingBoundaryCrossingRule boundaryCrossingRule)
 {
+    fprintf(stderr, "JDH: startCharacterOffsetOfParagraph\n");
```
This looks left over 😄

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -3758,8 +3758,10 @@ CharacterOffset AXObjectCache::previousBoundary(const CharacterOffset& character
     return characterOffsetForNodeAndOffset(*characterOffset.node, characterCount, TraverseOptionIncludeStart);
 }
 
+// MARK: replicate this!
```
This looks left over 😄

hoffmanjoshua replied:
Oops 🙈 thank you for catching this!

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,44 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+                // If a text run starts or ends with a newline character, that indicates a paragraph boundary.
+                // However, if the direction is Next, and our starting offset points to the end of the line (past the newline character),
+                // we are past the boundary.
+                if (currentRuns->at(i).endsWithLineBreak() && (i != startLineIndex || (direction == AXDirection::Next && currentRuns->runLength(i) != offsetInStartLine))) {
```
@twilco here is the new usage of `offsetInStartLine`, with a comment above explaining that we should continue searching when past the boundary.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -933,7 +933,44 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+            for (size_t i = startLineIndex; i < currentRuns->size() && i >= 0; direction == AXDirection::Next ? i++ : i--) {
+                // If a text run starts or ends with a newline character, that indicates a paragraph boundary.
+                // However, if the direction is Next, and our starting offset points to the end of the line (past the newline character),
+                // we are past the boundary.
```
But I bet we could make this comment fit on two lines if we move "However, if the direction" to the first line (may need to play with it a bit). Obviously not important though so don't worry if it doesn't work 😄

hoffmanjoshua committed:
AX: Implement TextUnit::Paragraph off the main thread
https://bugs.webkit.org/show_bug.cgi?id=282310
rdar://134579753

Reviewed by Tyler Wilcock.

This PR implements the current paragraph API to be served from the ax thread when AX_THREAD_TEXT_APIS
is enabled. Similar to computing the current line API, paragraph borders are determined by newline
characters, renderers with different containing blocks, and line break elements.

The text-marker-paragraph-nav.html test was partially ported to verify the behavior of this API.
A follow up PR will implement the previousParagraphStart and nextParagraphEnd APIs.

* LayoutTests/accessibility/ax-thread-text-apis/test.mp3: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-paragraph-nav.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::paragraphRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::previousParagraphStart const):
(WebCore::AXTextMarker::nextParagraphEnd const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRun::startsWithLineBreak const):
(WebCore::AXTextRun::endsWithLineBreak const):
(WebCore::AXTextRuns::runLength const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):

Canonical link: https://commits.webkit.org/286004@main
